### PR TITLE
[Tests] Add Common Test for Training + Fix a couple of bugs

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds an aggressive test to check that all models that should be trainable can perform a backward pass of their loss output. In addition, a test for training with gradient checkpointing is added as well. The motivation comes from this error: https://github.com/huggingface/transformers/pull/7562#issuecomment-723887221  - the PR introduced broke gradient checkpointing without any test noticing it. 

To make the test applicable for all models, some `ModelTests` have to overwrite the `_prepare_for_class` function.
In addition, some cleaning was done `ForPretraining` was renamed to `ForPreTraining`; an `AutoModelForNextSentencePredicition` was added, ...

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_albert.py [hunk 1]
```diff
@@ -749,7 +749,7 @@ def forward(
             Labels for computing the masked language modeling loss. Indices should be in ``[-100, 0, ...,
             config.vocab_size]`` (see ``input_ids`` docstring) Tokens with indices set to ``-100`` are ignored
             (masked), the loss is only computed for the tokens with labels in ``[0, ..., config.vocab_size]``
-        sentence_order_label (``torch.LongTensor`` of shape ``(batch_size,)``, `optional`):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_albert.py:752]: We only have two model types in the library that have a multi-loss model: 1) Some `ForPreTrainingModels` and the GPT `DoubleHeadModels`. I think it's cleaner if we force the loss names to be consistent here -> `sentence_order_label` is renamed to `next_sentence_label` as is done for `Bert`. The old argument is deprecated.

File: src/transformers/modeling_auto.py [hunk 2]
```diff
@@ -285,6 +289,7 @@
         (CTRLConfig, CTRLLMHeadModel),
         (ElectraConfig, ElectraForPreTraining),
         (LxmertConfig, LxmertForPreTraining),
+        (FunnelConfig, FunnelForPreTraining),
```
Comment (patrickvonplaten) [inline src/transformers/modeling_auto.py:292]: was missing

File: src/transformers/modeling_auto.py [hunk 3]
```diff
@@ -396,6 +401,7 @@
         (DebertaConfig, DebertaForSequenceClassification),
         (GPT2Config, GPT2ForSequenceClassification),
         (OpenAIGPTConfig, OpenAIGPTForSequenceClassification),
+        (ReformerConfig, ReformerForSequenceClassification),
```
Comment (patrickvonplaten) [inline src/transformers/modeling_auto.py:404]: was missing

File: src/transformers/modeling_auto.py [hunk 4]
```diff
@@ -417,6 +423,7 @@
         (ElectraConfig, ElectraForQuestionAnswering),
         (ReformerConfig, ReformerForQuestionAnswering),
         (FunnelConfig, FunnelForQuestionAnswering),
+        (LxmertConfig, LxmertForQuestionAnswering),
```
Comment (patrickvonplaten) [inline src/transformers/modeling_auto.py:426]: was missing

File: src/transformers/modeling_bert.py [hunk 5]
```diff
@@ -1228,13 +1228,14 @@ def forward(
         position_ids=None,
         head_mask=None,
         inputs_embeds=None,
-        next_sentence_label=None,
+        labels=None,
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:1231]: if model has only one loss as is the case here, I think we should force the name to be `labels`.

File: src/transformers/modeling_longformer.py [hunk 6]
```diff
@@ -1069,7 +1069,7 @@ def forward(
 
                 def create_custom_forward(module):
                     def custom_forward(*inputs):
-                        return module(*inputs, output_attentions)
+                        return module(*inputs, is_global_attn)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py:1072]: This fixes the longformer bug.

File: tests/test_modeling_albert.py [hunk 7]
```diff
@@ -227,6 +230,20 @@ class AlbertModelTest(ModelTesterMixin, unittest.TestCase):
         else ()
     )
 
+    # special case for ForPreTraining model
+    def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
```
Comment (patrickvonplaten) [inline tests/test_modeling_albert.py:234]: some models have this function added to their `...ModelTest`. It's necessary to make the `test_training` fn pass for all models.

File: tests/test_modeling_albert.py [hunk 8]
```diff
@@ -227,6 +230,20 @@ class AlbertModelTest(ModelTesterMixin, unittest.TestCase):
...
+        inputs_dict = super()._prepare_for_class(inputs_dict, model_class, return_labels=return_labels)
```
Comment (patrickvonplaten) [inline tests/test_modeling_albert.py:235]: since tests are not user-facing, I decided to use inheritance here instead of overwriting. Is that fine?

File: src/transformers/modeling_albert.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_albert.py:752]: Mmmm, this one has a different name than `next_sentence_label` because it is a different objective if my memory serves me right. So having a different name here does not seem like a bad idea.

File: src/transformers/modeling_auto.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/modeling_auto.py:292]: Good catch!

File: src/transformers/modeling_auto.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_auto.py:404]: Another good catch!

File: src/transformers/modeling_auto.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/modeling_auto.py:426]: We definitely need some kind of script to check those automatically ;-)

File: src/transformers/modeling_bert.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:1231]: I agree with that approach. This might require special new dataloaders to change the labels name, but this is easy enough to do.

File: tests/test_modeling_albert.py [hunk 9]
```diff
@@ -227,6 +230,20 @@ class AlbertModelTest(ModelTesterMixin, unittest.TestCase):
...
+
+        if return_labels:
+            if model_class in MODEL_FOR_PRETRAINING_MAPPING.values():
+                inputs_dict["labels"] = torch.zeros(
+                    (self.model_tester.batch_size, self.model_tester.seq_length), dtype=torch.long, device=torch_device
+                )
+                inputs_dict["next_sentence_label"] = torch.zeros(
```
Comment (sgugger) [inline tests/test_modeling_albert.py:242]: Might need some special case here for Albert if we keep it special label names.

File: tests/test_modeling_albert.py (hunk 8 above)
Comment (sgugger) [reply tests/test_modeling_albert.py:235]: No problem with this on my side.

File: tests/test_modeling_funnel.py [hunk 10]
```diff
@@ -360,6 +361,17 @@ class FunnelModelTest(ModelTesterMixin, unittest.TestCase):
         else ()
     )
 
+    # special case for ForPreTraining model
+    def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
+        inputs_dict = super()._prepare_for_class(inputs_dict, model_class, return_labels=return_labels)
+
+        if return_labels:
+            if model_class in MODEL_FOR_PRETRAINING_MAPPING.values():
```
Comment (sgugger) [inline tests/test_modeling_funnel.py:369]: Why not just test if its FunnelForPreTraining?

File: tests/test_modeling_electra.py [hunk 11]
```diff
@@ -285,6 +288,17 @@ class ElectraModelTest(ModelTesterMixin, unittest.TestCase):
         else ()
     )
 
+    # special case for ForPreTraining model
+    def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
+        inputs_dict = super()._prepare_for_class(inputs_dict, model_class, return_labels=return_labels)
+
+        if return_labels:
+            if model_class in MODEL_FOR_PRETRAINING_MAPPING.values():
```
Comment (sgugger) [inline tests/test_modeling_electra.py:296]: Why not just test if it's ElectraForPreTraining?

**Review by sgugger — APPROVED** [review]
This looks pretty cool! I'm not super sure about the change of label name for AlbertForPreTraining so let's check what @LysandreJik thinks of it. If we go through we it, there is one test to remove/adapt in test_data_collator.

File: src/transformers/modeling_albert.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/modeling_albert.py:752]: I think the SOP objective can still be represented by the `next_sentence_label`, as it can indicate whether the next sequence follows or precedes the first sequence.

However, `sentence_order_label` is a good name as it is, so why are we changing this? For consistency?

File: src/transformers/modeling_auto.py (hunk 4 above)
Comment (LysandreJik) [reply src/transformers/modeling_auto.py:426]: Yes, we do!

File: src/transformers/modeling_gpt2.py [hunk 12]
```diff
@@ -1069,7 +1069,7 @@ def forward(
             if self.num_labels == 1:
                 #  We are doing regression
                 loss_fct = MSELoss()
-                loss = loss_fct(pooled_logits.view(-1), labels.view(-1))
+                loss = loss_fct(pooled_logits.view(-1), labels.to(self.dtype).view(-1))
```
Comment (LysandreJik) [inline src/transformers/modeling_gpt2.py:1072]: :+1:

File: tests/test_modeling_albert.py (hunk 9 above)
Comment (LysandreJik) [reply tests/test_modeling_albert.py:242]: If it's still updated on a case-by-case basis, I don't really see the point in changing the name to `next_sentence_label`

**Review by LysandreJik — APPROVED** [review]
LGTM! Great changes, thanks a lot!

File: src/transformers/modeling_albert.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_albert.py]: Agree with you! Albert's Pretraining is slightly different to BERT's, so different names make sense here. Reverting the renaming.

Comment (patrickvonplaten) [reply src/transformers/modeling_albert.py]: Thought it's the exact same objective, but it's not
